### ISSUE TITLE
Remove concept of output workers

### ISSUE THREAD

**Issue by suyograo** [elastic/logstash#5653]
Output workers were introduced in LS 1.2(?) as a short-term solution to scale the synchronous communication between LS and ES, specifically for bulk indexing requests. The undesired side-effect was that output workers added another knob for users to tweak, and was often confused with `-w` flag -- "filter workers" or "pipeline workers", as its called now. There are 2 issues with output workers:

1. Make things simple, and remove extra knobs which only confuses users. These knobs are sometimes blindly tuned without fully understanding downstream side-effects. We should reduce the scope of guess work.
2. Output workers, while adding concurrency implies that outputs are inherently concurrent. It is uncertain what libraries Logstash uses is fully concurrent so it can be shared across multiple threads. In 2.2 we had prematurely increased output workers to be equal to pipeline workers which worked well in theory, but amplified existing concurrency issues in output plugins. This was later [reverted in 2.3](https://github.com/elastic/logstash/pull/4904) to go back to a default of 1 for output workers. For example, we got reports from [Elasticsearch Output](https://github.com/logstash-plugins/logstash-output-elasticsearch/issues/377) and [Kafka Output](https://github.com/elastic/logstash/issues/4754). Currently, Elasticsearch Output has been [rewritten](https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/390) in version 5.0 to be fully concurrent and to efficiently make use of resources like connections.

To summarize, we should standardize on pipeline workers (`-w`) as the only unit of scaling in Logstash. For this to happen, existing outputs must be audited to make sure underlying libraries are concurrent, thread safe and can effectively share resources. We should start with plugins packaged with Logstash by default and work our way from there. This is also related to auditing and making sure all plugins are reviewed for being "thread safe" w.r.t. using the `declare_threadsafe` semantic.

My suggestion is to deprecate `output_workers` in 5.0, and work to remove it in 6.0, given the above constraints are satisfied.

**Comment by djschny** [comment]
I'd like to have a discussion around keeping output workers in advocation that they are a good thing, not a workaround, and helpful for properly tuning something like Logstash.

Unless I'm missing something the fundamental issue of separating the pipeline works (filters essentially) from the output plugins is still important. How fast Logstash can process the filters is independent of the output factors. The pipeline filters/workers may be able to churn out work faster than the concurrent number of batch requests that we would want to be held open by the Elasticsearch output plugin. We don't want the Elasticsearch output plugin to automatically scale up based on the amount of data its getting from the pipeline or else we run into memory issues.
